internal/jobcontainers/job.go

```diff
@@ -0,0 +1,364 @@
+package jobcontainers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"unsafe"
+
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/winapi"
+	"golang.org/x/sys/windows"
+)
+
+// This file provides higher level constructs for the win32 job object API.
+// Most of the core creation and management functions are already present in "golang.org/x/sys/windows"
+// (CreateJobObject, AssignProcessToJobObject, etc.) as well as most of the limit information
+// structs and associated limit flags. Whatever is not present from the job object API
+// in golang.org/x/sys/windows is located in /internal/winapi.
+//
+// https://docs.microsoft.com/en-us/windows/win32/procthread/job-objects
+
+// jobObject is a high level wrapper around a Windows job object. Holds a handle to
+// the job and a handle to an iocp to be used to receive notifications about the lifecycle
+// of the job
+type jobObject struct {
+	jobHandle  windows.Handle
+	iocpHandle windows.Handle
+	handleLock sync.RWMutex
+}
+
+type jobLimits struct {
+	affinity       uintptr
+	cpuRate        uint32
+	cpuWeight      uint32
+	jobMemoryLimit uintptr
+	maxIops        int64
+	maxBandwidth   int64
+}
+
+var errAlreadyClosed = errors.New("the handle has already been closed")
+
+func (job *jobObject) setResourceLimits(ctx context.Context, limits *jobLimits) error {
+	if job.jobHandle == 0 {
+		return errAlreadyClosed
+	}
+	job.handleLock.RLock()
+	defer job.handleLock.RUnlock()
+
+	// Go through and check what limits were specified and construct the appropriate
+	// structs.
+	if limits.affinity != 0 || limits.jobMemoryLimit != 0 {
+		var (
+			basicLimitFlags uint32
+			basicInfo       windows.JOBOBJECT_BASIC_LIMIT_INFORMATION
+			eliInfo         windows.JOBOBJECT_EXTENDED_LIMIT_INFORMATION
+		)
+		eliInfo.BasicLimitInformation = basicInfo
+		if limits.affinity != 0 {
+			basicLimitFlags |= windows.JOB_OBJECT_LIMIT_AFFINITY
+			eliInfo.BasicLimitInformation.Affinity = limits.affinity
+		}
+		if limits.jobMemoryLimit != 0 {
+			basicLimitFlags |= windows.JOB_OBJECT_LIMIT_JOB_MEMORY
+			eliInfo.JobMemoryLimit = limits.jobMemoryLimit
+		}
+		_, err := windows.SetInformationJobObject(job.jobHandle, windows.JobObjectExtendedLimitInformation, uintptr(unsafe.Pointer(&eliInfo)), uint32(unsafe.Sizeof(eliInfo)))
+		if err != nil {
+			return fmt.Errorf("failed to set extended limit info on job object: %s", err)
+		}
+	}
+
+	if limits.cpuRate != 0 {
```

Comment:
Can we support `Weight` in addition to `CpuRate`?

Reply:
I had some oddities happening with this that are eluding me at the moment but I'll revisit

Reply:
Possibly because the internal type uses a union to represent cpurate and weight?

Reply:
I think you might want to update JOBOBJECT_CPU_RATE_CONTROL_INFORMATION to be 
```
type JOBOBJECT_CPU_RATE_CONTROL_INFORMATION struct {
	ControlFlags uint32
	Value uint32
}
```
Then you can use that value as either `CpuRate` or `Weight` or split it in two to get `MinRate` and `MaxRate` 
See: https://docs.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-jobobject_cpu_rate_control_information